### PR TITLE
issue-6628-android-support

### LINKED ISSUE #6730: Render-target with float format doesn't work
**Render-target with float format doesn't work**
I’ve done a number of experiments with float render-target with different formats.
This is the situation I have at the moment:
1. My build crashes when using ANY format other than R32F with error: ERROR:GRAPHICS: OpenGLSetTexture(2579): gl error 1280
2. R32F format works very strangely - there are no errors, but only if it is set to zero sampler (render.enable_texture(0, self.shadow_buffer).

**Expected behavior**
Here are some screenshots - one for PC, one for mobile. In these, I’ve displayed the contents of the shadow texture on the sprite on the right-hand side of the screen.
In addition, I’ve also displayed the content of the shadow texture also on top of the 3D objects, so that you can see how it’s projected.
This is screenshot for PC (we can see render-target projected on landscape and objects, similar as sprite at right screen side):
Скриншот 2022-06-19 11 10 11

This is screenshot for Android (we can see empty render-target on land, but correct result on sprite at right screen side):
Screenshot_20220619-110850
Why does render-target on Android look correct on sprite? The point is that zero sampler is used for the sprite, and everything works correctly on it.

This is a screenshot from Android. Here I've set the render-target texture to zero sampler and we can see that everything works correctly:
new

**Defold version:**
  - Version 1.3.3

**Platforms:**
 - Platforms: Android

**Minimal repro case project:**
[__SHADOW_TEST.zip](https://github.com/defold/defold/files/8943437/__SHADOW_TEST.zip)
controls:
- touch the left-hand side of the screen for move camera
- touch the right-hand side of the screen for rotate camera

### PR DESCRIPTION
**Pull request by Jhonnyg**
This is a fix for enabling MRT support on android. The problem is that we build and link with ES2 headers and library which doesn't have specifications for the functions needed to enable multiple outputs. This PR will load that symbol (glDrawBuffers) dynamically so we can use MRT where it is supported.

Fixes #6730

### CODE REVIEW

File: engine/render/src/render/render_script.cpp [hunk 1]
```diff
@@ -2722,9 +2722,13 @@ namespace dmRender
...
-        REGISTER_BUFFER_CONSTANT(COLOR1_BIT,  COLOR1_BIT);
-        REGISTER_BUFFER_CONSTANT(COLOR2_BIT,  COLOR2_BIT);
-        REGISTER_BUFFER_CONSTANT(COLOR3_BIT,  COLOR3_BIT);
+        // These depend on driver and context support
+        if (dmGraphics::IsMultiTargetRenderingSupported(graphics_context))
+        {
+            REGISTER_BUFFER_CONSTANT(COLOR1_BIT,  COLOR1_BIT);
+            REGISTER_BUFFER_CONSTANT(COLOR2_BIT,  COLOR2_BIT);
+            REGISTER_BUFFER_CONSTANT(COLOR3_BIT,  COLOR3_BIT);
+        }
```
Comment (Jhonnyg) [inline engine/render/src/render/render_script.cpp:2731]: This behaviour is similar to the floating point formats etc. If we have support for multiple outputs we add the enums

File: engine/graphics/src/opengl/graphics_opengl.cpp [hunk 2]
```diff
@@ -1054,6 +1063,8 @@ static uintptr_t GetExtProcAddress(const char* name, const char* extension_name,
             context->m_TextureFormatSupport |= 1 << TEXTURE_FORMAT_RG32F;
         }
 
+        context->m_IsMRTSupported = PFN_glDrawBuffers != 0x0;
```
[the remark this answers is not in the document]
Comment (JCash) [reply engine/graphics/src/opengl/graphics_opengl.cpp:1066]: Sounds good

File: engine/graphics/src/opengl/graphics_opengl.cpp [hunk 3]
```diff
@@ -2069,36 +2080,39 @@ static uintptr_t GetExtProcAddress(const char* name, const char* extension_name,
...
             {
-                uint32_t color_buffer_index = GetBufferTypeIndex(color_buffer_flags[i]);
-                rt->m_ColorBufferTexture[i] = NewTexture(context, creation_params[color_buffer_index]);
-                SetTexture(rt->m_ColorBufferTexture[i], params[color_buffer_index]);
+                uint32_t color_buffer_index = GetBufferTypeIndex(BUFFER_TYPE_COLOR0_BIT);
+                rt->m_ColorBufferTexture[0] = NewTexture(context, creation_params[color_buffer_index]);
+                SetTexture(rt->m_ColorBufferTexture[0], params[color_buffer_index]);
                 // attach the texture to FBO color attachment point
-                glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0 + i, GL_TEXTURE_2D, rt->m_ColorBufferTexture[i]->m_Texture, 0);
+                glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, rt->m_ColorBufferTexture[0]->m_Texture, 0);
```
Comment (JCash) [inline engine/graphics/src/opengl/graphics_opengl.cpp:2112]: I don't see a difference here versus the first loop iteration above?
Perhaps use:
```
uint32_t num_color_buffers = context->m_IsMRTSupported ? 4 : 1;
```

**Review by JCash — DISMISSED** [review]
Looks good
Small note on the for loop though

**Pull request merged**